rpcs3/evdev_joystick_handler.cpp

```diff
@@ -13,560 +17,755 @@
-                }
-
-                if (g_evdev_joystick_config.squirclejoysticks)
-                {
-                    joy_axis[i][axis] = value;
-                    if (evt.code == ABS_X || evt.code == ABS_Y)
-                    {
-                        int Xaxis = joy_axis_maps[i][ABS_X];
-                        int Yaxis = joy_axis_maps[i][ABS_Y];
-                        pad->m_sticks[Xaxis].m_value = scale_axis(ABS_X, joy_axis[i][Xaxis]);
-                        pad->m_sticks[Yaxis].m_value = scale_axis(ABS_Y, joy_axis[i][Yaxis]);
-
-                        std::tie(pad->m_sticks[Xaxis].m_value, pad->m_sticks[Yaxis].m_value) =
-                            ConvertToSquirclePoint(pad->m_sticks[Xaxis].m_value, pad->m_sticks[Yaxis].m_value);
-                    }
-                    else
-                    {
-                        int Xaxis = joy_axis_maps[i][ABS_RX];
-                        int Yaxis = joy_axis_maps[i][ABS_RY];
-                        pad->m_sticks[Xaxis].m_value = scale_axis(ABS_RX, joy_axis[i][Xaxis]);
-                        pad->m_sticks[Yaxis].m_value = scale_axis(ABS_RY, joy_axis[i][Yaxis]);
-
-                        std::tie(pad->m_sticks[Xaxis].m_value, pad->m_sticks[Yaxis].m_value) =
-                            ConvertToSquirclePoint(pad->m_sticks[Xaxis].m_value, pad->m_sticks[Yaxis].m_value);
-                    }
-                }
-                else
-                    pad->m_sticks[axis].m_value = scale_axis(evt.code, value);
-            }
-            break;
-        }
-        default:
-            LOG_ERROR(GENERAL, "Unknown joystick #%d event %d", i, evt.type);
-            break;
-        }
-    }
+	update_devs();
+
+	for (auto& device : devices)
+	{
+		auto pad = device.pad;
+		auto axis_orientations = device.axis_orientations;
+		auto& dev = device.device;
+		if (dev == nullptr) continue;
+
+		// Try to query the latest event from the joystick.
+		input_event evt;
+		int ret = libevdev_next_event(dev, LIBEVDEV_READ_FLAG_NORMAL, &evt);
+
+		// Grab any pending sync event.
+		if (ret == LIBEVDEV_READ_STATUS_SYNC)
+		{
+			LOG_NOTICE(GENERAL, "Captured sync event");
+			ret = libevdev_next_event(dev, LIBEVDEV_READ_FLAG_NORMAL | LIBEVDEV_READ_FLAG_SYNC, &evt);
+		}
+
+		if (ret < 0)
+		{
+			// -EAGAIN signifies no available events, not an actual *error*.
+			if (ret != -EAGAIN)
+				LOG_ERROR(GENERAL, "Failed to read latest event from joystick: %s [errno %d]", strerror(-ret), -ret);
+			continue;
+		}
+
+		bool is_negative = false;
+		int value;
+		int button_code = GetButtonInfo(evt, dev, value, is_negative);
+		if (button_code < 0 || value < 0)
+			continue;
+
+		bool is_button_or_trigger = evt.type == EV_KEY || button_code == ABS_Z || button_code == ABS_RZ;
+
+		// Translate any corresponding keycodes to our normal DS3 buttons and triggers
+		for (int i = 0; i < static_cast<int>(pad->m_buttons.size()); i++)
+		{
+			if (pad->m_buttons[i].m_keyCode != button_code)
```

Comment:
this seems problematic to me: pad->m_buttons[17].m_keyCode has a value of 0. So if I press on the ABS_Y (value of 1), it never gets to the line 656. If I press ABS_X (value 0), it enters the condition on line 656.